week-1-ts-ecosystem/w01e04-light-me-up/src/main.ts

```diff
@@ -9,13 +9,59 @@
  * */
 
 class PowerSource {
-  private energySupply = 100
+    private energySupply = 100
 
-  consume(energy) {
-    this.energySupply -= energy
-  }
+    consume(energy: number) {
+        if ((this.energySupply - energy) < 0) {
+            throw new Error('I\'m out of power')
+        }
+
+        this.energySupply -= energy
+    }
+
+    availableEnergy() {
+        return this.energySupply;
+    }
 }
 
 class LightBulb {
-  protected readonly powerConsumption = 20
+    protected readonly powerConsumption = 20
+    powerSource: PowerSource;
+
+    constructor(powerSource: PowerSource) {
+        this.powerSource = powerSource;
+    }
+
+    consume() {
+        this.powerSource.consume(this.powerConsumption);
+    }
+
+    lightOn(sec: number) {
+        for (let s = 0; s < sec; s++) {
+            setTimeout(() => {
+                this.consume();
+                console.log('Light on');
+            }, 1000);
+        }
+    }
 }
+
+/* Task 1 */
+
+const powerSource = new PowerSource();
+let secondsOn = 0;
+const interval = setInterval(() => {
+    try {
+        new LightBulb(powerSource).consume();
+        new LightBulb(powerSource).consume();
+        secondsOn++;
+    } catch (e) {
+        clearInterval(interval);
+        console.log(`Żarówki świeciły ${secondsOn} sek.`);
+    }
+}, 1000);
```

Comment:
Tutaj zasymulowałeś sobie 2 żarówki w czasie `n` sekund, i znów wkręcasz 2 nowe bazując na tym samym `power Source`. Pomysł - może być. Widać że czaisz o co chodzi (w zadaniu było żeby zapalić 6 żarówek ale ty to tym interwałem robisz - więc jest ok.).

Generalnie jest ok. poza szczegółem że jeśli któraś z nich rzuci błędem (w sensie `powerSource`) to nie doliczy to ostatniej sekundy bo linia `:57` nie wywoła się jeśli będzie błąd rzucony w `:55` lub `:56`.
Nie wiem czy nie bezpieczniej zliczać sekundy "przed" zapalaniem żarówek.